torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -64,6 +65,8 @@ void DistEngine::validateRootsAndRetrieveEdges(
       rootEdges, grads, [](const std::string& msg) { return msg; });
 }
 
+// NB: this function modifies the autograd graph. More specifically it replaces
```

Comment:
Note that since the graph could be used in other places, and in particular it could be used with the vanilla engine, you want to be careful here.
You should either forbid `retain_graph` to make sure no one will run the graph again or you should make this properly work in this case.

Reply:
I think it should work for `retain_graph = true`. I'll test that.

Reply:
In particular if you do mix the vanilla engine and the distributed one.